### PR TITLE
Creates a Rating Widget React Component

### PR DESCRIPTION
**Pull request by rsahlstrom**
Adds a react rating widget component that allows you to specify a size,
    control hover state, assign an initial rating, supports half stars, and
    can be disabled. Component also has a onRate callback that is called
    when a rating is assigned.

### CODE REVIEW

File: src/components/RatingWidget.jsx [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+var RatingWidget = React.createClass({
+  propTypes: {
+    size: React.PropTypes.number,
+    onRate: React.PropTypes.func,
+    disabled: React.PropTypes.bool,
+    initialRating: React.PropTypes.number,
+    halfRating: React.PropTypes.bool,
+    hover: React.PropTypes.bool,
+  },
```
Comment (rsahlstrom) [inline src/components/RatingWidget.jsx:16]: Here we're defining our propTypes so we can enforce what data we're expecting and can alert our implementers on what errors they're making.

File: src/components/RatingWidget.jsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      halfRating: false,
+      hover: true,
+    }
+  },
+
+  getInitialState: function() {
+    return {
+      rating: this.props.initialRating,
+      tempRating: null
+    }
```
Comment (rsahlstrom) [inline src/components/RatingWidget.jsx:36]: We use state to keep track of what the current rating is and what the mouseOver rating is.

File: src/components/RatingWidget.jsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    newRating = this.calchalfRating(newRating, e);
+    if (newRating === this.state.rating) {
+      newRating = 0;
+    }
+
+    this.setState({rating: newRating});
+    this.setState({tempRating: null});
+    this.props.onRate(newRating);
+  },
```
Comment (rsahlstrom) [inline src/components/RatingWidget.jsx:60]: All handleClick does is set the state. We do no DOM manipulation.

File: src/components/RatingWidget.jsx [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+    newTempRating = this.calchalfRating(newTempRating, e);
+    this.setState({tempRating: newTempRating})
+  },
+
+  handleOnMouseLeave: function() {
+    this.setState({tempRating: null});
+  },
+
+  calchalfRating: function(newRating, e) {
```
Comment (rsahlstrom) [inline src/components/RatingWidget.jsx]: I should probably rename this function to calcHalfRating. I'll fix that.

File: src/components/RatingWidget.jsx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    if (this.props.disabled) {
+      return;
+    }
+
+    if (!this.props.hover) {
+      return;
+    }
+
+    newTempRating = this.calchalfRating(newTempRating, e);
+    this.setState({tempRating: newTempRating})
```
Comment (rsahlstrom) [inline src/components/RatingWidget.jsx:68]: As above, hover only worries about setting state.

File: src/components/RatingWidget.jsx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        roundRating == ceilRating &&
+        this.props.halfRating
+      ;
+
+      var classes = {
+        'ddm-rating-widget__step': true,
+        'ddm-rating-widget__step--whole': showWhole,
+        'ddm-rating-widget__step--half': showHalf,
+        'ddm-rating-widget__step--hover': this.state.tempRating
+      }
```
Comment (rsahlstrom) [inline src/components/RatingWidget.jsx:133]: This is where the magic happens. We look at the state to determine which classes to add or remove from the step classes.

**Comment by rsahlstrom** [comment]
@kbratko Please review this and your other pull requests and then take it apart to see why we did what. Let me know any questions you have.

**Comment by kbratko** [comment]
After reviewing the code I would like to add some additional functionality for the best ui.
1.  Add cursor:pointer property to css class .ddm-rating-widget__step to show user that the rating can be modified. Also it would be better to add .ddm-rating-widget__step--disabled class when rating disabled property is true. And set up css property for this class cursor:default to show user that this rating can't be modified.
2. When I checked how rating works on my iPhone simulator I noticed that the stars is too small for being clicked correctly. It will be so difficult to tap them with fat fingers.
   So I decided to increase stars size. To change star size means that I should change .ddm-rating-widget__step font-size property. But in this case I should also change the height and width of it, which should be equal to the font-size (in case width  property has not correct value, the half star has wrong view). Also adding the line-height property, which also equal the font-size, allows to see the perfect square area around the star. 

So as we use sass (thats so great!) we can do it without any effort:

$stepSize: 30px;

.ddm-rating-widget__step {
  font-size: $stepSize;
  width: $stepSize;
  height: $stepSize;
  line-height: $stepSize;
}

or use mixin (make styles more reusable):

@mixin size($stepSize) {
  font-size: $stepSize;
  width: $stepSize;
  height: $stepSize;
  line-height: $stepSize;
}

.ddm-rating-widget__step { @include size(30px); }

It allows developer to change size in the only one place and all another properties can be set up automaticaly.

**Comment by kbratko** [comment]
One more thing. When I tap stars on iPhone the stars become blue and don't become orange when rate is done. As I know hover event doesn't work on mobile. Can we left orange color for both hover and whole states?

**Comment by rsahlstrom** [comment]
@kbratko Good feedback. I've pushed another commit with updated changes. Will you please take a look again?

@jspencerDDM Made some changes and got the disabled stuff working I tried yesterday. You may be interested in seeing this second commit.

File: src/components/RatingWidget.jsx (hunk 3 above)
Comment (jehoshua02) [inline src/components/RatingWidget.jsx:60]: I like to put it this way:

If a method modifies state, modifying state is the only thing it does.

This goes along with the [SOLID Single Responsibility Principle](http://en.wikipedia.org/wiki/SOLID_%28object-oriented_design). I like functions that only do one thing.

**Pull request merged**